onnxruntime/test/providers/kernel_def_hash_test.cc

```diff
@@ -0,0 +1,139 @@
+ */
+
+#include <algorithm>
+#include <cinttypes>
+#include <fstream>
+#include <iostream>
+
+#include "gtest/gtest.h"
+
+#ifdef _WIN32
+#pragma warning(push)
+#pragma warning(disable : 28020)
+#endif
+#include "nlohmann/json.hpp"
+#ifdef _WIN32
+#pragma warning(pop)
+#endif
+
+#include "asserts.h"
+#include "core/common/common.h"
+#include "core/common/path_string.h"
+#include "core/framework/kernel_registry.h"
+#include "core/mlas/inc/mlas.h"
+#include "core/providers/cpu/cpu_execution_provider.h"
+
+using json = nlohmann::json;
+
+namespace onnxruntime {
+namespace test {
+
+namespace {
+std::string DumpKernelDefHashes(const onnxruntime::KernelDefHashes& kernel_def_hashes) {
+  const json j(kernel_def_hashes);
+  return j.dump(/* indent */ 4);
+}
+
+KernelDefHashes ParseKernelDefHashes(std::istream& in) {
+  KernelDefHashes kernel_def_hashes{};
+  const json j = json::parse(in);
+  j.get_to<onnxruntime::KernelDefHashes>(kernel_def_hashes);
+  return kernel_def_hashes;
+}
+
+KernelDefHashes ReadKernelDefHashesFromFile(const PathString& path) {
+  std::ifstream in{path};
+  ORT_ENFORCE(in, "Failed to open file: ", ToMBString(path));
+  const auto kernel_def_hashes = ParseKernelDefHashes(in);
+  return kernel_def_hashes;
+}
+
+void CheckKernelDefHashes(const KernelDefHashes& actual, const KernelDefHashes& expected) {
+  ASSERT_TRUE(std::is_sorted(actual.begin(), actual.end()));
+  ASSERT_TRUE(std::is_sorted(expected.begin(), expected.end()));
+
+  constexpr const char* kNoteReference = "Note: Please read the note at the top of this file: " __FILE__;
+
+  KernelDefHashes expected_minus_actual{};
+  std::set_difference(expected.begin(), expected.end(), actual.begin(), actual.end(),
+                      std::back_inserter(expected_minus_actual));
+  EXPECT_TRUE(expected_minus_actual.empty())
+      << "Some expected kernel def hashes were not found.\n"
+      << kNoteReference << "\n"
+      << DumpKernelDefHashes(expected_minus_actual);
+
+  KernelDefHashes actual_minus_expected{};
+  std::set_difference(actual.begin(), actual.end(), expected.begin(), expected.end(),
+                      std::back_inserter(actual_minus_expected));
+  EXPECT_TRUE(actual_minus_expected.empty())
+      << "Unexpected kernel def hashes were found, please update the expected values as needed "
+         "(see the output below).\n"
+      << kNoteReference << "\n"
+      << DumpKernelDefHashes(actual_minus_expected);
+}
+}  // namespace
+
+TEST(KernelDefHashTest, DISABLED_PrintCpuKernelDefHashes) {
+  KernelRegistry kernel_registry{};
+  ASSERT_STATUS_OK(RegisterCPUKernels(kernel_registry));
+  const auto cpu_kernel_def_hashes = kernel_registry.ExportKernelDefHashes();
+  std::cout << DumpKernelDefHashes(cpu_kernel_def_hashes) << "\n";
+}
+
+// disabled by default because not all builds have the expected kernels
+// run with --gtest_also_run_disabled_tests
+TEST(KernelDefHashTest, DISABLED_ExpectedCpuKernelDefHashes) {
+  KernelRegistry kernel_registry{};
```

Comment:
It would be preferrable it this isn't disabled, as a developer won't see the failure until they run a CI (at it's not as simple to get the test output and fix the issue with that setup).

Can we instead use `#if !defined(REDUCED_OPS_BUILD)` to disable this in reduced builds only?

Reply:
it could be enabled. but then either we would need to have a more specific ifdef to only allow the build configuration that includes the right kernels (i.e., with training ops, contrib ops including nchwc ops, and ml ops), or have different expected lists for each case.

---
In reply to: [602936766](https://github.com/microsoft/onnxruntime/pull/7120#discussion_r602936766) [](ancestors = 602936766)